sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java

```diff
@@ -0,0 +1,131 @@
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.expression.scalar.timestamp;
+
+import io.crate.data.Input;
+import io.crate.expression.scalar.ScalarFunctionModule;
+import io.crate.metadata.BaseFunctionResolver;
+import io.crate.metadata.FunctionIdent;
+import io.crate.metadata.FunctionImplementation;
+import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.params.FuncParams;
+import io.crate.metadata.functions.params.Param;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import io.crate.types.LongType;
+import io.crate.types.TimestampType;
+
+import java.time.DateTimeException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Locale;
+
+public class TimezoneFunction extends Scalar<Long, Object> {
+
+    public static final String NAME = "timezone";
+    private static final ZoneId UTC = ZoneId.of("UTC");
+    private static final FuncParams FUNCTION_PARAMS = FuncParams.builder(Param.STRING, Param.ANY).build();
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(NAME, new BaseFunctionResolver(FUNCTION_PARAMS) {
+            @Override
+            public FunctionImplementation getForTypes(List<DataType> signatureTypes) throws IllegalArgumentException {
+                DataType returnType = null;
+                boolean paramHadTimezone = false;
+                int timestampParamTypeId = signatureTypes.get(1).id();
+                switch (timestampParamTypeId) {
+                    case LongType.ID:
+                    case TimestampType.ID_WITHOUT_TZ:
+                        returnType = DataTypes.TIMESTAMPZ;
+                        break;
+                    case TimestampType.ID_WITH_TZ:
+                        returnType = DataTypes.TIMESTAMP;
+                        paramHadTimezone = true;
+                        break;
+                    default:
+                        throw new IllegalArgumentException(String.format(
+                            Locale.ENGLISH,
+                            "invalid argument to %s: expected second parameter" +
+                            " to be of either type [%s, %s, %s] but found %s",
+                            NAME,
+                            LongType.INSTANCE.getName(),
+                            TimestampType.INSTANCE_WITH_TZ.getName(),
+                            TimestampType.INSTANCE_WITHOUT_TZ.getName(),
+                            signatureTypes.get(1)));
+                }
+                return new TimezoneFunction(new FunctionInfo(new FunctionIdent(NAME, signatureTypes), returnType),
+                                            paramHadTimezone);
```

Comment:
Would it make sense to infer this based on the `FunctionInfo` return type, instead of adding an extra parameter?

Reply:
You are absolutely right, I made this mistake because my brain was running low on stamina.